File: src/main/java/net/clgd/ccemux/emulation/CustomRomMount.java
package net.clgd.ccemux.emulation;

import static dan200.computercraft.core.filesystem.FileSystem.getDirectory;
import static dan200.computercraft.core.filesystem.FileSystem.getName;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

import org.apache.commons.io.IOUtils;

import dan200.computercraft.api.filesystem.IMount;

public class CustomRomMount implements IMount {

	private static String trimSlash(String path) {
		return path.replaceAll("/$", "");
	}
	
	private final Map<String, Map<String, byte[]>> folders;
	
	public CustomRomMount(ZipInputStream is) {
		folders = new HashMap<>();
		
		folders.put("", new HashMap<>()); // root folder
		
		try {
			ZipEntry entry;
			while ((entry = is.getNextEntry()) != null) {
				if (entry.isDirectory()) {
					folders.put(trimSlash(entry.getName()), new HashMap<>());
				} else {
					String n = trimSlash(entry.getName());
					byte[] data = new byte[(int)entry.getSize()];
					IOUtils.read(is, data);
					folders.get(getDirectory(n)).put(getName(n), data);
				}
			}
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	
	@Override
	public boolean exists(String path) {
		return folders.containsKey(path)
				|| folders.containsKey(getDirectory(path)) && folders.get(getDirectory(path)).containsKey(getName(path));
	}
	
	@Override
	public long getSize(String path) {
		if (folders.containsKey(getDirectory(path)) && folders.get(getDirectory(path)).containsKey(getName(path))) {
			return folders.get(getDirectory(path)).get(getName(path)).length;
		} else {
			return 0;
		}
	}
	
	@Override
	public boolean isDirectory(String path) {
		return folders.containsKey(path);
	}
	
	@Override
	public void list(String path, List<String> contents) {
		if (folders.containsKey(path)) {
			folders.get(path).keySet().forEach(f -> {
				if (!contents.contains(f)) contents.add(f);
			});
		}
	}
	
	@Override
	public InputStream openForRead(String path) {
		if (folders.containsKey(getDirectory(path))
				&& folders.get(getDirectory(path)).containsKey(getName(path))) {
			return new ByteArrayInputStream(folders.get(getDirectory(path)).get(getName(path)));
		} else {
			return null;
		}
	}
}

File: src/main/kotlin/net/clgd/ccemux/emulation/CustomRomMount.kt
package net.clgd.ccemux.emulation

import dan200.computercraft.api.filesystem.IMount
import dan200.computercraft.core.filesystem.FileSystem
import net.clgd.ccemux.Utils.trimSlashes
import org.apache.commons.io.IOUtils
import java.io.ByteArrayInputStream
import java.io.InputStream
import java.util.*
import java.util.zip.ZipInputStream

// here be dragons
class CustomRomMount(i: ZipInputStream) : IMount {
	val folders = HashMap<String, MutableMap<String, ByteArray>>()

	init {
		folders.put("", hashMapOf())

		while (true) {
			val entry = i.nextEntry ?: break

			if (entry.isDirectory) {
				folders.put(entry.name.trimSlashes(), hashMapOf())
			} else {
				val n = entry.name.trimSlashes()
				val data = ByteArray(entry.size.toInt())
				IOUtils.read(i, data)
				folders[FileSystem.getDirectory(n)]?.put(FileSystem.getName(n), data)
			}
		}
	}

	override fun exists(path: String): Boolean {
		return folders.containsKey(path) ||
			   folders[FileSystem.getDirectory(path)]?.containsKey(FileSystem.getName(path)) ?: false
	}

	override fun getSize(path: String?): Long {
		return folders.get(FileSystem.getDirectory(path))?.get(FileSystem.getName(path))?.size?.toLong() ?: 0L
	}

	override fun isDirectory(path: String) = folders.containsKey(path)

	override fun list(path: String, contents: MutableList<String>?) {
		folders.get(path)?.keys?.forEach { f -> if (!(contents?.contains(f) ?: true)) contents?.add(f) }

		folders.keys.forEach {
			it -> run {
				var f = it

				do {
					if (FileSystem.getDirectory(f) == path) {
						val n = FileSystem.getName(f)
						if (!(contents?.contains(n) ?: true)) contents?.add(n)
					}

					f = FileSystem.getDirectory(f)
				} while (f != "" && f != "..")
			}
		}
	}

	override fun openForRead(path: String): InputStream {
		return ByteArrayInputStream(folders.get(FileSystem.getDirectory(path))?.get(FileSystem.getName(path)))
	}
}
